private/buf/bufcheck/internal/bufcheckserver/internal/bufcheckserverhandle/lint.go

```diff
@@ -190,6 +197,1022 @@ func handleLintDirectorySamePackage(
+			file.PackageLocation(),
+			nil,
+			"Package name %q should be lower_snake.case, such as %q.",
+			pkg,
+			expectedPkg,
+		)
+	}
+	return nil
+}
+
+// HandleLintPackageNoImportCycle is a handle function.
+var HandleLintPackageNoImportCycle = bufcheckserverutil.NewLintFilesRuleHandler(handleLintPackageNoImportCycle)
+
+func handleLintPackageNoImportCycle(
+	responseWriter bufcheckserverutil.ResponseWriter,
+	_ bufcheckserverutil.Request,
+	files []bufprotosource.File,
+) error {
+	packageToDirectlyImportedPackageToFileImports, err := bufprotosource.PackageToDirectlyImportedPackageToFileImports(files...)
+	if err != nil {
+		return err
+	}
+	// This is way more algorithmically complex than it needs to be.
+	//
+	// We're doing a DFS starting at each package. What we should do is start from any package,
+	// do the DFS and keep track of the packages hit, and then don't ever do DFS from a given
+	// package twice. The problem is is that with the current janky package -> direct -> file imports
+	// setup, we would then end up with error messages like "import cycle: a -> b -> c -> b", and
+	// attach the error message to a file with package a, and we want to just print "b -> c -> b".
+	// So to get this to market, we just do a DFS from each package.
+	//
+	// This may prove to be too expensive but early testing say it is not so far.
+	for pkg := range packageToDirectlyImportedPackageToFileImports {
+		// Can equal "" per the function signature of PackageToDirectlyImportedPackageToFileImports
+		if pkg == "" {
+			continue
+		}
+		// Go one deep in the potential import cycle so that we can get the file imports
+		// we want to potentially attach errors to.
+		//
+		// We know that pkg is never equal to directlyImportedPackage due to the signature
+		// of PackageToDirectlyImportedPackageToFileImports.
+		for directlyImportedPackage, fileImports := range packageToDirectlyImportedPackageToFileImports[pkg] {
+			// Can equal "" per the function signature of PackageToDirectlyImportedPackageToFileImports
+			if directlyImportedPackage == "" {
+				continue
+			}
+			if importCycle := getImportCycleIfExists(
+				directlyImportedPackage,
+				packageToDirectlyImportedPackageToFileImports,
+				map[string]struct{}{
+					pkg: {},
+				},
+				[]string{
+					pkg,
+				},
+			); len(importCycle) > 0 {
+				for _, fileImport := range fileImports {
+					// We used newFilesWithImportsCheckFunc, meaning that we did not skip imports.
+					// We do not want to report errors on imports.
+					if fileImport.File().IsImport() {
+						continue
+					}
+					responseWriter.AddProtosourceAnnotation(
+						fileImport.Location(),
+						nil,
+						`Package import cycle: %s`,
+						strings.Join(importCycle, ` -> `),
+					)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// HandleLintPackageSameDirectory is a handle function.
+var HandleLintPackageSameDirectory = bufcheckserverutil.NewLintPackageToFilesRuleHandler(handleLintPackageSameDirectory)
```

Comment:
The original function calls [`newPackageToFilesCheckFunc`](https://github.com/bufbuild/buf/blob/4f697079bebbe88dc6b780bf8773f9046eaa90d8/private/bufpkg/bufcheck/buflint/internal/buflintcheck/util.go#L147), which includes imports, but `NewLintPackageToFilesRuleHandler` does not include import. So this check's behavior has deviated. Or we could add a `NewLintPackageToFilesIncludeImportRuleHandler`.

Same comment for directory to files.